File: content/examples/other/leetcode/problems/add-and-search-word/example-2/Solution.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
  static class TrieNode {
    Map<Character, TrieNode> children = new HashMap<>();
    boolean isWord = false;
  }

  static class WordDictionaryTrie {
    private final TrieNode root = new TrieNode();

    void addWord(String word) {
      TrieNode cur = root;
      for (int i = 0; i < word.length(); i += 1) {
        char ch = word.charAt(i);
        cur.children.putIfAbsent(ch, new TrieNode());
        cur = cur.children.get(ch);
      }
      cur.isWord = true;
    }

    boolean search(String word) {
      return dfs(root, word, 0);
    }

    boolean dfs(TrieNode node, String word, int i) {
      if (i == word.length()) return node.isWord;
      char ch = word.charAt(i);
      if (ch == '.') {
        for (TrieNode child : node.children.values()) {
          if (dfs(child, word, i + 1)) return true;
        }
        return false;
      }
      TrieNode next = node.children.get(ch);
      if (next == null) return false;
      return dfs(next, word, i + 1);
    }
  }

  public static void main(String[] args) {
    WordDictionaryTrie wd = new WordDictionaryTrie();
    wd.addWord("bad");
    wd.addWord("dad");
    wd.addWord("mad");
    System.out.println(wd.search("pad"));
    System.out.println(wd.search("bad"));
    System.out.println(wd.search(".ad"));
    System.out.println(wd.search("b.."));
  }
}

File: content/examples/other/leetcode/problems/add-and-search-word/example-2/solution.py
class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_word = False


class WordDictionaryTrie:
    def __init__(self):
        self.root = TrieNode()

    def add_word(self, word):
        cur = self.root
        for ch in word:
            if ch not in cur.children:
                cur.children[ch] = TrieNode()
            cur = cur.children[ch]
        cur.is_word = True

    def search(self, word):
        def dfs(node, i):
            if i == len(word):
                return node.is_word
            ch = word[i]
            if ch == ".":
                for child in node.children.values():
                    if dfs(child, i + 1):
                        return True
                return False
            if ch not in node.children:
                return False
            return dfs(node.children[ch], i + 1)

        return dfs(self.root, 0)


if __name__ == "__main__":
    wd = WordDictionaryTrie()
    wd.add_word("bad")
    wd.add_word("dad")
    wd.add_word("mad")
    print(wd.search("pad"))
    print(wd.search("bad"))
    print(wd.search(".ad"))
    print(wd.search("b.."))
